Freecadinit

Peque opened the pull request:
A bit of refactoring in FreeCAD's `__init__.py` file (making it a bit shorter and PEP8-compliant, as well as Python3-compatible).

Also added a new possible route for FreeCAD's library path (in Fedora this can be found under `lib64` instead of `lib`).

jmwright commented:
@Peque Thank you for the contributions, we really appreciate it. PEP8 cleanup and Python 3 updating have been on my mind for CadQuery for quite awhile. I do have a couple of questions about this PR.

1. Why did you convert the if-else statements for the operating system check to just if statements?
2. Why did you remove the try-except check when trying to import FreeCAD?

I'm not being critical, just wondering what the logic was.

Peque commented:
@jmwright Thanks for your feedback. :blush:

1. I fixed that in a new commit, it was a mistake.
2. Now the function `_fc_path()` is already rising the `ImportError` exception in case it does not return a valid `_PATH` (that removes the need of the `if path: [...] else: raise ImportError` part). Then I removed the last `try: import FreeCAD except: raise ImportError` because it would be actually hiding the import error in case it fails (if cadquery was really unable to find freecad library path, then the error would have been raised before, when calling `_fc_path()`.

Please, tell me if you see anything else or if you want me to change anything (I can revert `2.` if you don't like how it is implemented now).

Talking about Python 3 support, I'm having a look at [yorik's FreeCAD Python changes](https://github.com/yorikvanhavre/FreeCAD) and I will try to contact him to check the state of their Python 3 support. If it is working right now (even if it is not merged yet), I may work on making cadquery's code Python 3-compatible (and, on the way, I may help a bit with PEP8 changes). I think it would be a good way to dive into cadquery's code (as I would like to use this library).

jmwright commented:
@dcowden I'm ready to merge this unless you have anything to add.

@Peque  Awesome, thanks for doing that. I understand your logic on #2 now, thanks. Good catch.

I have been watching the FreeCAD-Python 3 progress off and on. I thought I had a better link than this, but it's all I found when I did a quick search just now. http://www.freecadweb.org/tracker/view.php?id=995

If I launch freecad-daily from the Ubuntu PPA I see that it's still using Python 2.7.12, but I wouldn't be surprised if the FreeCAD team is getting close to making the switch to Python 3.

We have a community Google Group if you're interested in joining it: https://groups.google.com/forum/#!forum/cadquery

You can ask general questions, share CadQuery scripts you've written, etc. We can start a Python 3 update discussion if you're interested in that.

Thanks for jumping on board. We're happy to help you get started, so if you need help feel free to ask.

Here are the main docs too:
http://dcowden.github.io/cadquery/

Peque commented:
It seems there is actually some work going on with the Python 3 port. There are very recent updates in [the FreeCAD forum thread](http://forum.freecadweb.org/viewtopic.php?f=10&t=12534). Also, yorik [mentioned in his blog this september](http://yorik.uncreated.net/guestblog.php?2016=256):

> [...] But there is much more to come. FreeCAD is being ported to new verisons of Qt (Qt5) and Python (python3). Most of the work is done, and we might soon see that included. [...]

Maybe those changes are not in the Ubuntu PPA, only in his fork. I will investigate it further and try it myself to see how it goes.

Thanks for the links. :blush:

Peque commented:
@jmwright I had a look at yorik's FreeCAD Python 3 branch, but wasn't able to compile. The work is stalled right now, although he says he will be working on it again soon. However, there is still more work to do: FreeCAD 0.16 now depends on VTK (version 6), which support only Python 2. Python 3 support is available in VTK 7 apparently, but not tested yet. That means in order to have FreeCAD working with Python 3, we may need to wait several months (I don't see that happening in the next 4 months).

I could dedicate my time to push that change, and then perhaps work on a Conda distributable FreeCAD package so that it can be easily used by cadquery. However, I don't feel too motivated to do that.

On the other hand, I saw there was some work/intention on cadquery to implement an OCC backend, and eliminate the FreeCAD dependency. Is there any progress on that? I would rather help to push that forward than to push FreeCAD's Python 3 support.

jmwright commented:
@Peque Thanks for checking on that. It's great to have an update so that we know what to expect.

@dcowden started working on the OCC back end again in 2016. I believe his current branch is [here](https://github.com/dcowden/cadquery/tree/2_0_branch). The readme should give you a good idea of where Dave is headed with that version. At some point in 2017 the current FreeCAD-based version of CadQuery is going to be bumped up to v1.0.x, which would make the number convention of v2.0.x for the PythonOCC version make more sense.

I also started on an Electron/Web based front end for CadQuery that would run with either the FreeCAD or PythonOCC based versions. Work on that has stalled for now though. That codebase is [here](https://github.com/jmwright/cadquery-gui) if you're interested.

Peque commented:
@jmwright Thanks. The OCC branch you pointed me to seems to be empty (i.e.: nothing really implemented).

@dcowden is there a more up-to-date branch where I could find an initial OCC implementation? (even if very basic or with missing functionality, but at least with something working?)

jmwright commented:
@Peque Here's a branch that I experimented with about a year ago. The problem I tan into is that it's extremely hard to do a direct port of the current CadQuery to PythonOCC. https://github.com/dcowden/cadquery/tree/pythonocc?files=1

Here's an older branch of Dave's. https://github.com/dcowden/cadquery/tree/pythonOCC-backend?files=1

I'm not sure how useful those are at this point.

I'll let Dave confirm, but I think you are correct. The newest OCC branch is a skeleton that Dave set up the way he thinks 2.0 should be structured.

dcowden commented:
@Peque, yes @jmwright is correct. I started the process of sorting to pure
pythonOCC but it is a lot of work. Freedcad has quite. Lot of high level
code that's simply a chore to reproduce.

I abandoned the work not because it is impossible. But rather because I
found that other work, such as dxf import and easier installation would be
much more valuable.

All that said, I would absolutely love to get progress on a pure OCC
version. If you are interested in contributing , I would love to help! I
have a decent amount of experience with the OCC Api, so I can provide some
guidance.

On Nov 30, 2016 7:54 PM, "Jeremy Wright" <notifications@github.com> wrote:

> @Peque <https://github.com/Peque> Here's a branch that I experimented
> with about a year ago. The problem I tan into is that it's extremely hard
> to do a direct port of the current CadQuery to PythonOCC.
> https://github.com/dcowden/cadquery/tree/pythonocc?files=1
>
> Here's an older branch of Dave's. https://github.com/dcowden/
> cadquery/tree/pythonOCC-backend?files=1
>
> I'm not sure how useful those are at this point.
>
> I'll let Dave confirm, but I think you are correct. The newest OCC branch
> is a skeleton that Dave set up the way he thinks 2.0 should be structured.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dcowden/cadquery/pull/157#issuecomment-264045928>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABPOA9KfkoWiGxu0gIsn4l_tZ5y2qgoVks5rDhqsgaJpZM4K9QMN>
> .
>

jmwright commented:
@Peque You mentioned wanting to dive in and learn how to use CadQuery. There are lots of ways to do that depending on exactly what you want to get out of it. If you want to dive in to the deep end on OCC that's great and we'll support you the best we can, but if that's too much to take on there are other options. Here are a few off the top of my head.

1. Working through the examples embedded with the CadQuery module for FreeCAD, updating and adding comments as needed.
2. Taking over an issue(s) in the tracker. We have several open issues including one bug that bothers me, but that I haven't had time for. I think it involves a bug at either the FreeCAD or OCC level that's giving us trouble, so there's some deep work that probably needs to happen to fix it. https://github.com/dcowden/cadquery/issues/83
There are also several feature requests in the issue tracker as well. https://github.com/dcowden/cadquery/issues
3. Doing a sweep of Python 2-3 crossover conversion and PEP8 cleanup in the code. This might get a little tedious though since it's not as in-depth as a full Python 3 update.
4. Start working on something you're interested in creating in CadQuery on the Google Group so we can follow along and help as needed.

So there are multiple options to get involved depending on what your interests are and your end goal with CadQuery. Let us know if any of the options I've outlined, or any others above (like the PythonOCC implementation), interest you.

dcowden commented:
Thanks @jmwright!

I have some time available over the holidays. During that time, I'm going to spend some time on DXF imports ( in collaboration with @huskier ).

I will also update the roadmap. Based on input from @huskier, I'm moving DXF work ahead of docker/conda.

@Peque, we'll take all the help we can get.  Choose something you'd enjoy working on, and we'll support you.  The only requirements we have on accepting PRs is that test  coverage doesnt go the wrong direction.

Peque commented:
Thanks @jmwright and @dcowden for your kindness.

Among those things, I would only really be interested in working on a Python CAD library, with a "pythonic" API and with does not have FreeCAD as a dependency. So that would mean working with the Python OCC backend for cadquery, yes. However, as you mention, that might be a hard work, specially if there is not so much done yet. So I have to give it a thought, and decide whether I am interested in investing a decent amount of time into that...

I will probably play a little bit more with FreeCAD, cadquery and other tools to learn about the current state of CAD development in Python and then, after holidays (in January), decide. After all, I have just begun to create my first 3D designs and I still need to learn a bit more about this world. :blush:

By the way, what do you think about this projects?

- https://github.com/floatingpointstack/aoc-utils
- http://tenko.github.io/occmodel/index.html

Do you think they could be used as a starting point or reference? Did you already now about them?

dcowden commented:
Hi @Peque 

I didn't know about either of those modules. I can only assume that occmodel would be a thinner ( and thus more even more frustrating to use ) interface than pythonOCC.

I have known Thomas over at https://github.com/tpaviot/pythonocc for a while-- i always presumed that would be the basis. They do have conda packages to make isntallation of OCC+OCE relatively plainless.

jmwright commented:
@Peque Thanks for the links, I had not seen those. It would be interesting to create the same semi-complex model in PythonOCC, PythonOCC/aoc-utils, and occmodel to see how they compare.

I created a [script](https://github.com/jmwright/pythonocc_oce_setup) to compile and install OCE (the community edition of OCC) and PythonOCC on Ubuntu 14.04. I haven't tried it yet, but I suspect it will also work on 16.04. That should help you get OCC up and running quickly, if you haven't already.

Dave and I have had numerous conversations about the pros and cons of having CadQuery based on FreeCAD. I'm curious why you would prefer working with an API that avoids having FreeCAD in the stack. I don't have an agenda in asking that, I'm just genuinely interested in why users would prefer one route over the other.

Peque commented:
@dcowden Yes, I had already found their conda packages. I find them very convenient. :blush:

@jmwright I bet your decision is well informed, in my case it is probably just a feeling. I feel FreeCAD is a huge dependency to be kept as a back-end for cadquery (i.e.: does too many things not needed by cadquery, and distribution/installation is also not very convenient for a Python package). On the other hand, PythonOCC feels like a better (more concise/restricted) back-end for cadquery.

adam-urbanczyk commented:
@Peque I am also interested in working on OCC back-end. Could you join the cadquery group to discuss https://groups.google.com/forum/#!forum/cadquery ?

jmwright commented:
@Peque Here's a direct link to the current discussion on the OCC implementation. https://groups.google.com/forum/m/#!msg/cadquery/1O-_DHPmIDE/vEM7o5UVCAAJ

kmolLin commented:
Will cadquery update new version in pthon3 version, use cadquery is easier to build model than pythonocc.

jmwright commented:
Hi @kmolLin

We are planning to support Python 3, but there will probably be a period of time where CadQuery is broken on FreeCAD while we're fixing conversion problems.

adam-urbanczyk commented:
@kmolLin 

If you feel adventurous you might take a look here: 

https://github.com/adam-urbanczyk/cadquery

It is an experimental CQ implementation based on OCC. Seems to be working (with some exceptions) on py3.

kmolLin commented:
@jmwright  
Thanks to Reply
I want to use cadquery in Parametric design not in freecad.
It can build the 3D model with step or iges.

kmolLin commented:
Hi @adam-urbanczyk  
I will try to use the version in your fork repository.
Thank's reply my question.
